### ISSUE TITLE
Strides in caffe wrappers for Standard BLAS functions 

### ISSUE THREAD

**Issue by venkai** [BVLC/caffe#5249]
I was wondering why the stride option is not allowed in any of the caffe BLAS wrappers.
All functions in math_functions.hpp assume a default stride of 1.
This can easily be modified without breaking existing code by taking a default stride argument = 1.
Is there a particular reason why the stride option is disallowed ? 

To be more precise, is non-contiguous memory access detrimental to gpu performance ?

**Comment by naibaf7** [comment]
@venkai Yes non-contiguous memory access can be detrimental, but there are ways to implement a BLAS with strides not equal to 1 and not loose too much performance.
The reason it is not included is just that there was no need to use it in Caffe so far. If you need it to write a particular layer, just change it.

**Comment by venkai** [comment]
@naibaf7 I agree with whatever you said. However, I still fail to understand why strides are not part of the argument list. It is very easy to modify any function in math_functions.hpp to have a default stride argument (=1). However, in my opinion, it seems that the authors of caffe have intentionally decided to drop strided BLAS/memcpy operations for a particular reason.

As a simple example, consider **A=W*D**; where **D**=diag(**_d_**); Aside from using loops, you can do this:
1) loop over indices in **_d_**, scale column _i_ in  **W** by **d**[_i_]; 
2) loop over rows _j_ in **W**, assign **A**(_j_,:) = _hadamard_(**W**(_j_,:),**_d_**);

Option 1) seems more intuitive and is also faster.
Option 2) is better that 1) _only_ when strides are unavailable.
